[Candidate_parameters] Improve front-end validation for entering Consent data

zaliqarosli opened the pull request:
### Brief summary of changes
This PR makes entering consent data more rigorous. Going forward, this PR implements the following workflow for data entry:

- Date of consent being given is required for both answers to consent "yes" or "no"
- Date of withdrawal of consent required if answer changes from "yes" to "no"
- Consent status "yes" or "no" cannot be changed to an empty one
- Replaces confirmation message to swal
- Adds script to check errors in data

### To test this change...

- [ ] Make sure you have consent set up (https://github.com/aces/Loris/blob/minor/modules/candidate_parameters/setup.md)
- [ ] In Candidate Parameters -> Consent Status tab:

1. Date of Withdrawal field should be disabled until it is needed.
2. For a currently empty consent status field, change status to "yes". Date of Consent should be required.
3. For a currently empty consent status field, change status to "no". Date of Consent should be required. Date of withdrawal still disabled.
4. For a consent status field with a "yes" status, change answer to "no". Date of Withdrawal should be enabled and required. Save answer.
5. Change the same status field back to "yes". Date of Withdrawal is disabled and the date cleared.
6. For a consent status field with a "no" or "yes" status, you cannot change it back to empty. 

**Note for reviewing: While there are plenty of pre-ES6 syntax, conversion of vars to lets/consts does not belong in this PR. All new code highlighted in green should already conform to ES6 standards.**

me-evans commented:
As discussed in today's LORIS meeting, weighing in re: proposed consent changes.  This is great, just have a few minor adjustments to request.  

- Date of consent being given is required for both answers to consent "yes" and "no"
    - Agree completely
- Date of withdrawal of consent required if answer changes from "yes" to "no"
    - Agree, just propose also blocking them from :
        - Entering Date of withdrawal as the first date with either Yes or No (To prevent errors and enforce a proper consent history --> if they're behind and only get to entering consent once it's been withdrawn, I think we should still require that an entry of 'Yes' with the 'Date of Consent' be saved before allowing a new entry with the 'No' / 'Date of Withdrawal' combo
        - Entering a 'Date of Withdrawal' if the consent is still 'Yes'
- If "yes" or "no" answer to consent already exists, the current status cannot be erased to an empty answer
    - Agree with @LeighMac that some users should have this permission (same people that would be allowed to edit candidate parameters that have been saved should be allowed to remove entries made by mistake).  We could make it very explicit so it is only used when appropriate.  For example:
        - Have a separate button for clearing an entry made by mistake that would produce a pop-up warning that they should only proceed if the entry was made by mistake (like wrong candidate) and reiterating what to do if they just want to update instead of erase/force them to click 'proceed' or 'dismiss'

zaliqarosli commented:
Changes made :

1. Date of Withdrawal field is disabled by default. It is enabled when either :

- consent is being withdrawn (status changes from "yes" to "no") OR
- date of withdrawal already exists (allows user to remove/change date)

2. Date of Withdrawal field is required when:

- consent is being withdrawn (status changes from "yes" to "no")

3. Removed disallowing users to erase a valid status as the ideal workflow drawn out above requires more work to be done. A GitHub issue can be made to address this @me-evans . Please review code to double check the logic. I added comments to make the code human readable.

zaliqarosli commented on modules/candidate_parameters/jsx/ConsentStatus.js:
```diff
@@ -100,18 +100,30 @@ var ConsentStatus = React.createClass(
-            withdrawalRequired[i] = false;
+          // If answer to consent is "no", require date of consent
+          if (this.state.formData[consent] === "no") {
+            dateRequired[i] = true;
```
may as well!

ridz1208 commented:
@zaliqarosli just tested this, minor changes needed.

- I cant seem to be able to clear the date of withdrawal and save
- whenthe consent changes from yes/no to empty, the date fields should all be cleared.

zaliqarosli commented:
PR description updated to address Rida's comments above

ridz1208 commented:
@zaliqarosli still not working

ridz1208 commented:
@zaliqarosli new rules need to be applied on the backend through a release script

zaliqarosli commented:
@ridz1208 please re-review!

ridz1208 commented on tools/single_use/Cleanup_Consent_Data.php:
```diff
@@ -0,0 +1,48 @@
+ */
+require_once __DIR__ . '/../generic_includes.php';
+
+$db     = \NDB_Factory::singleton()->database();
```
you should probably be using the databse in generic includes

zaliqarosli commented:
@ridz1208 done. please re-review

zaliqarosli commented on tools/single_use/Cleanup_Consent_Data.php:
```diff
@@ -0,0 +1,48 @@
+            break;
+    }
+}
+print_r($errors);
```
@done. please re-review

ridz1208 replied:
@zaliqarosli who's @done ??

zaliqarosli replied:
hahaha i meant "@ridz1208 done" 😆
